web/server/go.md

```diff
@@ -0,0 +1,498 @@
+
+#### Working with Models in Go
+
+In the second case,
+we need to utilize our models to store results of our application logic.
+After all, model structs and databases don't deliver client value on their own.
+Let's take a look back at our dog model
+and perform some operations.
+
+First, let's try to add a dog.
+
+```go
+package models
+
+type Dog struct {
+    ID      uuid.UUID `json:"id" db:"id"`
+    Name    string    `json:"name" db:"name"`
+    Owner     Owner
+}
+```
+
+Elsewhere in the code,
+we create a new dog and save it.
+
+```go
+package services
+
+func MyDogFunc(name string, owner Owner) (Dog, error) {
+    dog := Dog{}
+    dog.Name = name
+    dog.Owner = Owner
+    // Save is an external method to interface with DB passed as dependency
+    db.Save(&dog)
+    return dog, nil
+}
+```
+
+Notice how our data source is queried entirely outside of the model.
+This allows us to separate the model structure
+and the dependencies such as external packages for DB connectors
+or the type of database itself.
+Using this separation helps to isolate data source specific code,
+allowing mutability of underlying infrastructure
+and enforcing code reuse.
+
+Let's say our Owner can leave the application,
+and in this case the owner's dogs should also be marked as inactive.
+We want our Dog model code to maintain this constraint.
+
+Here's a method to demonstrate that relationship:
+
+```go
+package models
+
+type Dog struct {
+    // ...
+    Status Status
+}
+
+type Owner struct {
+    Dogs []Dog
+    Status Status
+}
+
+func (o *Owner) Deactivate() {
+    o.Status = Inactive
+    dogs := o.Dogs[:0]
+    for _, dog := range o.Dogs {
+        dog.Status = Inactive
+        dogs = append(dogs, dog)
+    }
+    o.Dogs = dogs
+}
+```
+
+The purpose here is to set up the public model API
+to only allow valid modifications throughout the codebase.
+We've explicitly linked the dog and owner
+to enforce any data constraints.
+Similar methodology may follow for reactivating the owner,
+changing a dog's owner, etc.
+Note that in many cases when writing Go code,
+we'd set parameters like these as unexported
+to avoid invalid state changes,
+however this isn't possible due to required exposure
+by DB/JSON packages that will use the model's public fields.
+
+We've continued to separate any DB concerns
+to outside the model.
+For example,
+it's likely we'd want to wrap these changes in a transaction:
+
+```go
+package services
+
+func DeactiveOwner(owner Owner) error {
+    transaction := db.Transaction.Begin()
+    owner.Deactivate()
+    err := transaction.Save(owner)
+    if err !=nil {
+        transaction.Rollback()
+        return err
+    }
+    err = transaction.Save(owner.Dogs)
+    if err !=nil {
+        transaction.Rollback()
+        return err
+    }
+    err := transaction.Save(owner)
+    transaction.Commit()
+    return nil
+}
+```
+
+Separating the database concerns from the model
```

Comment:
This especially stems for a "concurrency first" approach. Writing DB code/queries into the model makes them so inflexible. I understand the pattern from inheritance based models (such as Active Record), but it doesn't seem worthwhile with the interface/composition based Go patterns. I especial see this with locking updates are introduced into a codebase. The second application logic gets spliced in the middle of that isolation, writing it into the models gets heavily overloaded. It's possible this is an anti-pattern and would love some input.

Reply:
I am not convinced that separating the model layer from the database provides enough benefits to justify the overhead of an additional layer. I think you're going to need integration tests that test the models and database interacting as long as you have any kind of constraints in the database, which means that separating the database and the models needs to provide enough benefits to justify now needing to write an additional set of tests for the models in isolation in addition to the overhead of managing another abstraction.

IMHO applications don't change databases often enough for that possibility to drive application architecture. I think we need better justification for this recommendation.

Reply:
I am _very_ in favor of keeping model and data access layers separate. (I deeply distrust the Active Record pattern.) This also matches the recommendations of the onion layer model of looking at code: https://www.codeguru.com/csharp/csharp/cs_misc/designtechniques/understanding-onion-architecture.html

Reply:
@jim what's the benefit of keeping these together? The models have to be passed via struct, since their fields are required for marshalling. They're also passed across multiple layers by necessity of data passing. This means all their methods are going to be passed as well and pulled in via package imports/exposed publicly. If the DB connector/interface is included, it is now essentially permanent/hard to move in the codebase due to its propagation. These seem like easy concerns to separate, since those methods aren't required to pass data.

I don't think it's common for someone to point blank change a DB per say (although, who knows if we'll work on some non-postgres projects with different data needs), but it is common to change the connector or access patterns. For example, applications moving from a single monolith talking directly to a DB to services with APIs require this sort of separation for a refactor without breaking their client contracts.